File: platforms/windows/src/FileID.App/Services/PerceptualGrouping.cs
// Perceptual near-duplicate grouping — pure + testable.
// The engine stores a 64-bit difference-hash (dHash) per image in `files.phash`
// (the Int64 bit-pattern of a UInt64). Two images are "visually similar" when the
// Hamming distance of their dHashes is small; union-find clusters any transitive
// chain of within-threshold pairs into one group.

using System;
using System.Collections.Generic;
using System.Numerics;
using System.Threading;

namespace FileID.Services;

internal static class PerceptualGrouping
{
    internal const long MaxCandidateComparisons = 25_000_000;

    public static int HammingDistance(long a, long b)
        => BitOperations.PopCount((ulong)a ^ (ulong)b);

    public static List<List<long>> GroupByHamming(
        IReadOnlyList<(long Id, long Phash)> items,
        int maxHamming,
        CancellationToken cancellationToken = default)
        => GroupByHammingMeasured(items, maxHamming, cancellationToken).Groups;

    internal static (List<List<long>> Groups, long Comparisons) GroupByHammingMeasured(
        IReadOnlyList<(long Id, long Phash)> items,
        int maxHamming,
        CancellationToken cancellationToken = default)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(maxHamming);
        ArgumentOutOfRangeException.ThrowIfGreaterThan(maxHamming, 64);

        int n = items.Count;
        var groups = new List<List<long>>();
        if (n <= 1) return (groups, 0);

        var parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;

        int Find(int x)
        {
            int root = x;
            while (parent[root] != root)
            {
                parent[root] = parent[parent[root]];
                root = parent[root];
            }
            return root;
        }

        void Union(int a, int b)
        {
            int rootA = Find(a);
            int rootB = Find(b);
            if (rootA == rootB) return;
            if (rootA < rootB) parent[rootB] = rootA;
            else parent[rootA] = rootB;
        }

        var unique = new List<int>(n);
        var representativeByHash = new Dictionary<long, int>(n);
        for (int i = 0; i < n; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (representativeByHash.TryGetValue(items[i].Phash, out var representative))
            {
                Union(representative, i);
            }
            else
            {
                representativeByHash.Add(items[i].Phash, i);
                unique.Add(i);
            }
        }

        long comparisons = 0;
        if (maxHamming >= 64)
        {
            for (int i = 1; i < unique.Count; i++) Union(unique[0], unique[i]);
        }
        else if (maxHamming > 0 && unique.Count > 1)
        {
            int blockCount = maxHamming + 1;
            var buckets = new Dictionary<(int Block, ulong Value), List<int>>(unique.Count * 2);
            foreach (var index in unique)
            {
                cancellationToken.ThrowIfCancellationRequested();
                ulong hash = (ulong)items[index].Phash;
                for (int block = 0; block < blockCount; block++)
                {
                    ulong value = BlockValue(hash, block, blockCount);
                    var key = (block, value);
                    if (!buckets.TryGetValue(key, out var candidates))
                    {
                        candidates = new List<int>();
                        buckets.Add(key, candidates);
                    }

                    foreach (var candidate in candidates)
                    {
                        if ((comparisons & 0xFFF) == 0)
                        {
                            cancellationToken.ThrowIfCancellationRequested();
                        }
                        ulong candidateHash = (ulong)items[candidate].Phash;
                        if (SharesEarlierBlock(hash, candidateHash, block, blockCount)) continue;
                        comparisons++;
                        if (comparisons > MaxCandidateComparisons)
                        {
                            throw new InvalidOperationException(
                                $"Visually similar comparison exceeded the safety budget of {MaxCandidateComparisons:N0} candidate pairs. Reduce FILEID_NEARDUP_HAMMING or use Exact cleanup.");
                        }
                        if (BitOperations.PopCount(hash ^ candidateHash) <= maxHamming)
                        {
                            Union(candidate, index);
                        }
                    }
                    candidates.Add(index);
                }
            }
        }

        var order = new List<int>();
        var membersByRoot = new Dictionary<int, List<long>>(n);
        for (int i = 0; i < n; i++)
        {
            int root = Find(i);
            if (!membersByRoot.TryGetValue(root, out var members))
            {
                members = new List<long>();
                membersByRoot[root] = members;
                order.Add(root);
            }
            members.Add(items[i].Id);
        }

        foreach (var root in order)
        {
            var ids = membersByRoot[root];
            if (ids.Count >= 2) groups.Add(ids);
        }
        return (groups, comparisons);
    }

    private static bool SharesEarlierBlock(
        ulong left,
        ulong right,
        int currentBlock,
        int blockCount)
    {
        for (int block = 0; block < currentBlock; block++)
        {
            if (BlockValue(left, block, blockCount) == BlockValue(right, block, blockCount))
            {
                return true;
            }
        }
        return false;
    }

    private static ulong BlockValue(ulong hash, int block, int blockCount)
    {
        int start = block * 64 / blockCount;
        int end = (block + 1) * 64 / blockCount;
        int width = end - start;
        ulong mask = width == 64 ? ulong.MaxValue : (1UL << width) - 1;
        return (hash >> start) & mask;
    }
}

File: platforms/apple/shared/Sources/FileIDShared/PerceptualGrouping.swift
// Perceptual near-duplicate grouping — pure + testable. The engine stores a
// 64-bit difference-hash (dHash) per image in `files.phash` (the Int64
// bit-pattern of a UInt64). Two images are "visually similar" when the Hamming
// distance of their dHashes is small; union-find clusters any transitive chain
// of within-threshold pairs into one group. Kept dependency-free so SharedTests
// can drive it directly, separate from the GRDB query that feeds it.
import Foundation

public enum PerceptualGrouping {
    /// Hamming distance of two 64-bit dHashes: popcount(a XOR b) over the raw
    /// 64 bits. We XOR the *bit patterns*, not the signed Int64 values — the
    /// sign bit is just bit 63 of the hash.
    public static func hammingDistance(_ a: Int64, _ b: Int64) -> Int {
        (UInt64(bitPattern: a) ^ UInt64(bitPattern: b)).nonzeroBitCount
    }

    /// Union-find clustering: items whose dHashes are within `maxHamming` of one
    /// another — transitively (A~B, B~C ⇒ {A,B,C}) — form a group. Returns groups
    /// of size >= 2, each as its member ids in first-seen order, with the groups
    /// themselves in first-seen order. O(N²) pairwise; callers guard input size.
    public static func groupByHamming(
        _ items: [(id: Int64, phash: Int64)],
        maxHamming: Int
    ) -> [[Int64]] {
        let n = items.count
        guard n > 1 else { return [] }

        var parent = Array(0..<n)
        func find(_ x: Int) -> Int {
            var r = x
            while parent[r] != r { parent[r] = parent[parent[r]]; r = parent[r] }
            return r
        }
        func union(_ a: Int, _ b: Int) {
            let ra = find(a), rb = find(b)
            // Point the higher index at the lower one so every component's root
            // is its smallest member index — keeps group order deterministic.
            if ra != rb { parent[max(ra, rb)] = min(ra, rb) }
        }

        for i in 0..<n {
            for j in (i + 1)..<n where hammingDistance(items[i].phash, items[j].phash) <= maxHamming {
                union(i, j)
            }
        }

        var order: [Int] = []
        var membersByRoot: [Int: [Int64]] = [:]
        for i in 0..<n {
            let r = find(i)
            if membersByRoot[r] == nil { order.append(r) }
            membersByRoot[r, default: []].append(items[i].id)
        }
        return order.compactMap { root in
            let ids = membersByRoot[root]!
            return ids.count >= 2 ? ids : nil
        }
    }
}
